### PR TITLE
Maas instance plugin

### PR DESCRIPTION
**Pull request by YujiOshima**
Support BareMetal deployment.
An Instance plugin for  [Canonical MaaS](https://maas.io/) node.
#398

### CODE REVIEW

**Comment by chungers** [comment]
@YujiOshima  - nice work.  Just a general question...  does MAAS support the notion of detachable disk / volumes?   At least with the AWS plugin at https://github.com/docker/infrakit.aws it understands the `Attachment` of type `ebs`.   This ability is very interesting for cases of supporting stateful instances where attachable volumes store state and instances can come and go.  

Does this model make sense for MAAS or for bare-metal where disks/volumes can be attached /detached from compute instances?

**Comment by YujiOshima** [comment]
@chungers Thank you!
That's good idea and important function for common "Pet" use case in the on-premises ! I didn't test that use case, but I think It can be done with iSCSI.
I will try to support volume store along with other functions. Now MaaS plugin is very incomplete, should be enhanced step by step.

**Comment by Lennie** [comment]
In the meantime you could always use Docker volume plugin. For example just have 1 NFS share and create sub directories in your container when they don't exist yet. That way you don't need to request a share.

**Comment by YujiOshima** [comment]
@Lennie Thank you for the comment. Only for docker swarm flavor and an app in the docker container, It is true. But for other flavor or outside of docker container (I don't know the specific use case, may be docker daemon config or something), you need Infrakit volume management.

**Comment by Lennie** [comment]
Just saying workarounds are possible if you need to deploy soon and can't wait. Or need something for a dev. env. on a laptop. Also just realized you can create NFS server as a docker container as well.

But would be great to have a proper option for MAAS. But also don't forget the dev. laptop.

**Comment by YujiOshima** [comment]
@Lennie Thank you for a good point. Yes, we need to do a lot of things before volume management.
At first I think that we can restrict the scope to only docker swarm or vanilla flavor.

File: examples/instance/maas/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+Default login and password: `admin` `pass`
+
+Import boot images in `Images` tab.
+
+! Wait for import images. It takes some time. !
+
+Set up nodes.
+
+```
+./setup-nodes.sh
```
Comment (chungers) [inline examples/instance/maas/README.md]: Is this file included in this PR?  I don't see it.   Would you mind including it?

File: examples/instance/maas/instance.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+	for k, v := range labels {
+		tags[k] = v
+	}
+
+	encoded, err := types.AnyValue(tags)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(tagFile, encoded.Bytes(), 0666)
```
Comment (chungers) [inline examples/instance/maas/instance.go]: Files are written locally to store tags...  is there a way in MAAS to tag resources and query by tags?  https://docs.ubuntu.com/maas/2.1/en/intro-concepts#tags  

If we can reduce amount of state we need store locally, the better.     If there are ways to do tagging via the MAAS api, then we should consider making that change in a future PR?

File: examples/instance/maas/instance.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+			if err := os.RemoveAll(machineDir); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// DescribeInstances returns descriptions of all instances matching all of the provided tags.
+func (m maasPlugin) DescribeInstances(tags map[string]string) ([]instance.Description, error) {
```
Comment (chungers) [inline examples/instance/maas/instance.go]: See comment above -- is there a way in MAAS to query by tags of the nodes?   Using files here is ok but we'd have to make sure these files are highly available.

File: examples/instance/maas/maas-vanilla.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (chungers) [inline examples/instance/maas/maas-vanilla.yml]: This file is named `.yml` but the content is a JSON...

File: examples/instance/maas/README.md (hunk 1 above)
Comment (YujiOshima) [inline examples/instance/maas/README.md]: I update the Readme. Now I have only a simple document, but I will enhance it with the addition of functions in the future.

File: examples/instance/maas/instance.go (hunk 2 above)
Comment (YujiOshima) [inline examples/instance/maas/instance.go]: I think that is a better implementation. This time we implemented it in a simple way that does not depend on the api version, but we will change the information to keep it in a robust way without duplication.

File: examples/instance/maas/maas-vanilla.yml (hunk 4 above)
Comment (YujiOshima) [inline examples/instance/maas/maas-vanilla.yml]: Oh..This is a mistake. Fixed.

File: examples/instance/maas/README.md (hunk 1 above)
Comment (chungers) [inline examples/instance/maas/README.md]: Sorry I was confused.  This is a file from your repo... I see now, so there's no need to include this.

**Comment by chungers** [comment]
@YujiOshima - Before I can merge this PR, can you please add a few tests?    You can look at the `plugin_test.go` in the [terraform plugin](../tree/master/examples/instance/terraform/) as an example?   

Also, once this PR is accepted, would you mind open an issue on using MAAS tagging API to implement the instance state persistence?   I'd want to make sure that if we accept this PR this plugin will continue to be improved and is useful for those learning to work with InfraKit.  Thank you!

**Comment by YujiOshima** [comment]
@chungers Yeah! I added a few tests!
> open an issue

Of course! I will do so soon. And I want to proceed while discussing.

**Comment by YujiOshima** [comment]
I used a gomaasapi.testservice for test this plugin.
You can test using this for the basic function of MasS.

**Pull request merged**